### PR TITLE
feat(wash-cli): Add --watch flag to wash app list for displaying live info

### PR DESCRIPTION
**Pull request by Aditya1404Sal**
## Feature or Problem
This flag perpetually displays changes to application status until Ctrl+c is invoked

## Related Issues
#2512

### CODE REVIEW

**Comment by Aditya1404Sal** [comment]
https://github.com/user-attachments/assets/c27351f4-7fb9-4573-a192-244783a352dc

Here's the progress for `wash app list --watch`

**Comment by Aditya1404Sal** [comment]
Update 2:

https://github.com/user-attachments/assets/d9ca6eae-d30f-4970-b1ed-cf440a354b9b

The `request timed out` error came after I ran `wash down -- all` and the output is 90% responsive
The live info adapts to the terminal size and it runs the same in the vscode terminal too

**Review by brooksmtownsend — CHANGES_REQUESTED** [review]
This worked pretty well for me 😄 

I had one supernit&request, but my real requested change here may be more of a personal preference. This implementation output the refreshed information every second, but what I expected was more of an implementation like `watch` where the screen is cleared & new information is written to the terminal.

Did you find this as an option when looking into crossterm, or give it a try and find issues? Curious on your thoughts there

**Comment by Aditya1404Sal** [comment]
> This worked pretty well for me 😄 
> 
> I had one supernit&request, but my real requested change here may be more of a personal preference. This implementation output the refreshed information every second, but what I expected was more of an implementation like `watch` where the screen is cleared & new information is written to the terminal.
> 
> Did you find this as an option when looking into crossterm, or give it a try and find issues? Curious on your thoughts there

By clearing the screen do you mean clearing out the history of the terminal and Displaying the table on a clean terminal or logging of new information, because the current implementation does clear the table and re-displays the info after a sleep duration.
I tweaked `watch_application` in a way that keeps the terminal history intact instead of clearing it out.

Do let me know if you'd prefer the former or correct me if I'm wrong of my understanding.


At first I tried the alternate terminal method but the output was disoriented and the labels weren't properly spaced each time, at first I thought it must've been cause of zsh , but then I tried it with sh and bash , which gave the same results

File: crates/wash-cli/src/app/mod.rs [hunk 1]
```diff
@@ -152,8 +161,13 @@ pub async fn handle_command(
...
         List(cmd) => {
-            sp.update_spinner_message("Listing applications ...".to_string());
-            get_applications(cmd).await?
+            if cmd.watch {
+                sp.finish_and_clear();
+                get_applications(cmd).await?
+            } else {
+                sp.update_spinner_message("Listing applications ...".to_string());
+                get_applications(cmd).await?
+            }
```
[the remark this answers is not in the document]
Comment (Aditya1404Sal) [reply crates/wash-cli/src/app/mod.rs:170]: Sure thing!

File: crates/wash-cli/src/app/mod.rs [hunk 2]
```diff
@@ -50,6 +57,8 @@ pub enum AppCliCommand {
 pub struct ListCommand {
     #[clap(flatten)]
     opts: CliConnectionOpts,
+    #[clap(long)]
+    watch: bool,
```
[the remark this answers is not in the document]
Comment (Aditya1404Sal) [reply crates/wash-cli/src/app/mod.rs:61]: will do!

File: crates/wash-cli/src/app/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aditya1404Sal) [reply crates/wash-cli/src/app/mod.rs:170]: So like this ?
```rust
        List(cmd) => {
            sp.update_spinner_message("Listing applications ...".to_string());
            if cmd.watch {
                sp.finish_and_clear();
            }
            get_applications(cmd).await?
        }
```

[the remark this answers is not in the document]
Comment (brooksmtownsend) [reply crates/wash-cli/src/app/mod.rs]: I was thinking to just remove the change here and instead `finish_and_clear` down in `get_applications` when you're doing the `if cmd.watch` check, this is just a nit to cut down on the code branching so it's optional if that doesn't make a difference really

[the remark this answers is not in the document]
Comment (Aditya1404Sal) [reply crates/wash-cli/src/app/mod.rs]: Ah now I understand, will change that 👍🏻

**Comment by brooksmtownsend** [comment]
@Aditya1404Sal here's what it looks like for me, it seems like in an empty terminal or a mostly empty terminal it does the re-clearing, but sometimes it just continually outputs to the screen 🤔 This is made a little bit worse with tmux

https://github.com/user-attachments/assets/2af0ea28-12ed-45f5-b5ac-3150e067ef6c

Is there anything you can think of that would cause this behavior? What I see in the beginning of the video is exactly what I wanted, and I'm totally okay with clearing the previous output like you were doing before if it redraws in the same place

**Comment by Aditya1404Sal** [comment]
> @Aditya1404Sal here's what it looks like for me, it seems like in an empty terminal or a mostly empty terminal it does the re-clearing, but sometimes it just continually outputs to the screen 🤔 This is made a little bit worse with tmux
> 
>  wash.mov 
> Is there anything you can think of that would cause this behavior? What I see in the beginning of the video is exactly what I wanted, and I'm totally okay with clearing the previous output like you were doing before if it redraws in the same place

Yeah I have noticed it bugging out very rarely, I think it has something to do with the cursor tracking when the terminal isn't  in vscode but i'm not too sure.
I'll be happy to proceed with the previous version of clearing out the entire history and then redrawing in the same place.

**Comment by brooksmtownsend** [comment]
> I'll be happy to proceed with the previous version of clearing out the entire history and then redrawing in the same place.

@Aditya1404Sal I think that will work well 👍🏻 let's give that a try, and if you can test it out in another terminal I'll give it a test when you push up as well.

**Comment by Aditya1404Sal** [comment]
@brooksmtownsend I've modified the code to clear the entire terminal, The issue of multiple tables being printed out is gone (as far as i've tested) However the terminal if resized on specific intervals gives strange output, the table get's disoriented if the terminal's width is changed while the command runs.

**Comment by brooksmtownsend** [comment]
>  However the terminal if resized on specific intervals gives strange output, 

I think this is fine, there's really nothing we can do to handle folks resizing their terminal a bunch. I don't think it's too much to ask to re-run a command if you do this

**Review by brooksmtownsend — APPROVED** [review]
LGTM!

**Comment by Aditya1404Sal** [comment]
> > However the terminal if resized on specific intervals gives strange output,
> 
> I think this is fine, there's really nothing we can do to handle folks resizing their terminal a bunch. I don't think it's too much to ask to re-run a command if you do this

I'll still try to find a solution for this issue :rocket:

**Pull request merged**